File: host/signal/protocol.go
// Package signal 实现 Host 侧的 WebSocket 信令端点与 WebRTC PeerConnection
// 编排(answerer 角色)。
//
// CONTEXT.md 词汇:
//   - Signal:Host 与 Client 之间经 WebSocket 交换 SDP offer/answer 与 ICE
//     candidate 的过程,只用于建连协商,不承载媒体或输入。
//   - Offerer/Answerer Roles:Client 永远是 SDP offerer,Host 永远是 answerer。
//   - Trickle ICE:候选边收集边经 signal 推送。
//   - Signal Origin Check:Host 在 WS 握手阶段校验 Origin header。
//
// 本 slice(#3)只建 PeerConnection,不挂任何 media track / datachannel
// (那些是 #4–#7)。Session ID 在此 slice mint 并回写给 Client;抢占 teardown
// 与 stale-input-drop 留给 #6/#8。
package signal

import (
	"encoding/json"
	"fmt"
)

// FrameType 是信令帧类型。wire format 是 JSON,每帧带 type 字段。
type FrameType string

const (
	// Client → Host
	FrameTypeAuth      FrameType = "auth"      // 握手后第一帧:{"type":"auth","password":"..."}
	FrameTypeOffer     FrameType = "offer"     // Client 主动发起的 SDP offer
	FrameTypeCandidate FrameType = "candidate" // trickle ICE 候选

	// Host → Client
	FrameTypeAuthOK FrameType = "auth_ok" // 认证通过 + 回写 Session ID
	FrameTypeAnswer FrameType = "answer"  // Host 回的 SDP answer
	// FrameTypeCandidate 复用(双向)
	FrameTypeError FrameType = "error" // 密码错 / Origin 拒绝 / 协议错
)

// Frame 是 WebSocket 上传输的信令帧统一信封。
//
// 不同 type 用不同字段:
//
//	auth      → Password
//	auth_ok   → SessionID
//	offer     → SDP(webrtc.SessionDescription.Serialize 的字符串形式)
//	answer    → SDP
//	candidate → Candidate(webrtc.ICECandidateInit 的 JSON 字符串)
//	error     → Message
type Frame struct {
	Type      FrameType `json:"type"`
	Password  string    `json:"password,omitempty"`
	SessionID int64     `json:"sessionId,omitempty"`
	SDP       string    `json:"sdp,omitempty"`
	Candidate string    `json:"candidate,omitempty"`
	Message   string    `json:"message,omitempty"`
}

// readFrame 从 JSON 字节解析一个 Frame。
func readFrame(data []byte) (Frame, error) {
	var f Frame
	if err := json.Unmarshal(data, &f); err != nil {
		return Frame{}, fmt.Errorf("signal: invalid frame json: %w", err)
	}
	return f, nil
}

// writeFrame 序列化 Frame 为 JSON 字节。
func writeFrame(f Frame) ([]byte, error) {
	return json.Marshal(f)
}

File: client/src/signal/protocol.ts
// Client 侧信令帧定义。与 Host 的 host/signal/protocol.go 严格对齐。
//
// wire format 是 JSON,每帧带 type 字段。详见 host/signal/protocol.go 的注释
// 与 CONTEXT.md 词汇:Signal / Offerer-Answerer / Trickle ICE / 单密码认证 /
// Signal Origin Check。

export type FrameType =
  | 'auth' // C→H:握手首帧,带 password
  | 'auth_ok' // H→C:认证通过,回写 sessionId
  | 'offer' // C→H:Client 主动发起的 SDP offer
  | 'answer' // H→C:Host 回的 SDP answer
  | 'candidate' // 双向:trickle ICE 候选
  | 'error' // H→C:密码错 / Origin 拒绝 / 协议错

export interface Frame {
  type: FrameType
  password?: string
  sessionId?: number
  /** SDP 帧:JSON 对象 {"type":"offer|answer","sdp":"<SDP 串>"} 的字符串化形式。 */
  sdp?: string
  /** candidate 帧:RTCICECandidateInit 的 JSON 字符串。 */
  candidate?: string
  message?: string
}

// SDPDescription 是 sdp 字段里塞的 JSON 对象。与浏览器 RTCSessionDescription
// 的 {type, sdp} 形状一致,可直接喂给 setLocalDescription / setRemoteDescription。
export interface SDPDescription {
  type: 'offer' | 'answer' | 'pranswer' | 'rollback'
  sdp: string
}

/** 序列化一个 SDPDescription 为字符串(放进 Frame.sdp)。 */
export function serializeSDP(desc: RTCSessionDescriptionInit | SDPDescription): string {
  return JSON.stringify({ type: desc.type, sdp: desc.sdp })
}

/** 反序列化 Frame.sdp 字段为 SDPDescription。 */
export function parseSDP(raw: string): SDPDescription {
  return JSON.parse(raw) as SDPDescription
}
